llvm/lib/Support/SpecialCaseList.cpp

```diff
@@ -113,22 +120,49 @@ bool SpecialCaseList::createInternal(const std::vector<std::string> &Paths,
       Error = (Twine("can't open file '") + Path + "': " + EC.message()).str();
       return false;
     }
-    std::string ParseError;
-    if (!parse(FileOrErr.get().get(), ParseError)) {
-      Error = (Twine("error parsing file '") + Path + "': " + ParseError).str();
+    if (!createInternal(FileOrErr->get(), Error)) {
+      Error = llvm::formatv("error parsing file '{0}': {1}", Path, Error);
       return false;
     }
   }
   return true;
 }
-
 bool SpecialCaseList::createInternal(const MemoryBuffer *MB,
                                      std::string &Error) {
-  if (!parse(MB, Error))
+  auto ParsedInput = ParsedSpecialCaseList::parse(*MB);
+  if (!ParsedInput) {
+    Error = llvm::toString(ParsedInput.takeError());
+    return false;
+  }
+  if (auto Err = mergeSections(*std::move(ParsedInput))) {
+    Error = llvm::toString(std::move(Err));
     return false;
+  }
   return true;
 }
 
+llvm::Error SpecialCaseList::mergeSections(ParsedSpecialCaseList ParsedInput) {
+  bool UseGlobs = !ParsedInput.UseRegexes;
+  for (auto &S : ParsedInput.Sections) {
```

Comment:
RE "auto", the type is not trivial from the local code, so perhaps specify the type?
If the type is too long, perhaps avoid inner structs.